### PR TITLE
add UpdateRuntimeConfig interface

### PR DESCRIPTION
**Pull request by freehan**
Expose UpdateRuntimeConfig interface in RuntimeService for kubelet to pass a set of configurations to runtime. Currently it only takes PodCIDR.

The use case is for kubelet to pass configs to runtime. Kubelet holds some config/information which runtime does not have, such as PodCIDR. I expect some of kubelet configurations will gradually move to runtime, but I believe cases like PodCIDR, which dynamically assigned by k8s master, need to stay for a while.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34276)

### CODE REVIEW

File: pkg/kubelet/api/services.go [hunk 1]
```diff
@@ -86,3 +86,9 @@ type ImageManagerService interface {
 	// RemoveImage removes the image.
 	RemoveImage(image *runtimeApi.ImageSpec) error
 }
+
+// RuntimeManagerService interface should be implemented by a container image manager.
+// The methods should be thread-safe.
+type RuntimeManagerService interface {
+	PushRuntimeConfig(config *runtimeApi.RuntimeConfig) error
```
Comment (yujuhong) [inline pkg/kubelet/api/services.go]: Shouldn't this just be part of the `RuntimeService`?

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 2]
```diff
@@ -676,3 +682,21 @@ message RemoveImageRequest {
...
+
+enum ConfigKey {
+    POD_CIDR = 1;
+    HAIRPIN_MODE = 2;
+    NON_MASQUERADE_CIDRS = 3;
+    NETWORK_PLUGIN_DIR = 4;
+}
+
+message RuntimeConfig {
+    required ConfigKey key = 1;
```
Comment (yujuhong) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: The convention is to not use any `required` fields, which has been removed in proto3.

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 3]
```diff
@@ -676,3 +682,21 @@ message RemoveImageRequest {
...
 message RemoveImageResponse {}
+
+enum ConfigKey {
+    POD_CIDR = 1;
+    HAIRPIN_MODE = 2;
+    NON_MASQUERADE_CIDRS = 3;
+    NETWORK_PLUGIN_DIR = 4;
+}
+
+message RuntimeConfig {
```
Comment (yujuhong) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: Is this `NetworkConfig` or a more generic `RuntimeConfig`? i think `PushRuntimeConfig` should be a general purpose call that can be used to update a single config object (not restricted to network).

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 4]
```diff
@@ -55,6 +55,12 @@ service ImageService {
     rpc RemoveImage(RemoveImageRequest) returns (RemoveImageResponse) {}
 }
 
+// Runtime manager service defines the public APIs for managing runtime
+service RuntimeManagerService {
+    // PUshRuntimeConfig pushes a RuntimeConfig to runtime.
```
Comment (yujuhong) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: nit: s/PUsh/Push

**Review by yujuhong — CHANGES_REQUESTED** [review]

**Comment by mrunalp** [comment]
:+1:

**Comment by feiskyer** [comment]
@freehan Will the pushed configs be changed over time? If not, I suggest rename it to `InitRuntimeConfig` and pass those configs to runtime at initialization stage.

**Comment by freehan** [comment]
@feiskyer 
PodCIDR is assigned by master after kubelet registers. It is not determined upon initialization. Hence need a way for kubelet to pass it down to runtime. I suspect there are other cases like this.

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 5]
```diff
@@ -676,3 +679,17 @@ message RemoveImageRequest {
 }
 
 message RemoveImageResponse {}
+
+message NetworkConfig {
+    optional string podCIDR = 1;
```
Comment (feiskyer) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: Please use CamelCase for field names and add doc comments for it, e.g. `pod_cidr`.

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 6]
```diff
@@ -676,3 +679,17 @@ message RemoveImageRequest {
...
+}
+
+message RuntimeConfig {
+    optional NetworkConfig networkConfig = 1;
```
Comment (feiskyer) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: same here, `network_config`

File: pkg/kubelet/api/v1alpha1/runtime/api.proto [hunk 7]
```diff
@@ -676,3 +679,17 @@ message RemoveImageRequest {
...
+}
+
+message UpdateRuntimeConfigRequest {
+    optional RuntimeConfig runtimeConfig = 1;
```
Comment (feiskyer) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: same here,`runtime_config`

**Comment by euank** [comment]
cc @squeed

File: pkg/kubelet/api/v1alpha1/runtime/api.proto (hunk 5 above)
Comment (freehan) [inline pkg/kubelet/api/v1alpha1/runtime/api.proto]: @squeed Currently, this is a one time update when podCIDR become available. 
Later if we support the concept of multiple networks, then we need to figure out if we want to pass that information at RunPodSandbox.

**Comment by freehan** [comment]
@k8s-bot test this

**Review by yujuhong — APPROVED** [review]

**Comment by yujuhong** [comment]
@freehan, `verify-flags-underscore.py` is failing
```
Verifying hack/make-rules/../../hack/verify-test-images.sh
[0;32mSUCCESS[0m  hack/make-rules/../../hack/verify-test-images.sh	0s
Verifying hack/make-rules/../../hack/verify-flags-underscore.py
Found illegal 'flag' usage. If these are false negatives you should run `hack/verify-flags-underscore.py -e > hack/verify-flags/exceptions.txt` to update the list.
pkg/kubelet/api/v1alpha1/runtime/api.pb.go:	PodCidr          *string `protobuf:"bytes,1,opt,name=pod_cidr,json=podCidr" json:"pod_cidr,omitempty"`
pkg/kubelet/api/v1alpha1/runtime/api.pb.go:	RuntimeConfig    *RuntimeConfig `protobuf:"bytes,1,opt,name=runtime_config,json=runtimeConfig" json:"runtime_config,omitempty"`
pkg/kubelet/api/v1alpha1/runtime/api.proto:    optional RuntimeConfig runtime_config = 1;
pkg/kubelet/api/v1alpha1/runtime/api.proto:    optional string pod_cidr = 1;
[0;31mFAILED[0m   hack/make-rules/../../hack/verify-flags-underscore.py	59s
Makefile:99: recipe for target 'verify' failed
```

**Comment by yujuhong** [comment]
LGTM. 

@euank @yifan-gu, anymore question?

**Comment by euank** [comment]
LGTM, though we probably could have also included the remote_runtime implementation in this PR without trouble. I assume that'll follow soon enough.

**Comment by freehan** [comment]
Tests passed. Applying lgtm based on comments.

**Pull request merged**